docs/tutorials/gallery_adding_components.py

```diff
@@ -0,0 +1,125 @@
+# or call a local LLM. Ragna is designed to give you that flexibility.
+
+# %%
+# Your [`answer()`][ragna.core.Assistant.answer] method should take a prompt in the form of a
+# string, and a list of [`Source`][ragna.core.Source]s, in addition to whatever other arguments
+# necessary for your particular assistant. The return type is an [`Iterator`](https://docs.python.org/3/library/stdtypes.html#typeiter) of strings.
+
+# %%
+# !!! note
+#     Ragna also supports streaming responses from the assistant. See the
+#     [example how to use streaming responses](../../generated/examples/gallery_streaming.md)
+#     for more information.
+
+# %%
+# ## Adding a Source Storage
+
+from ragna.core import Document, Source, SourceStorage
+
+
+class TutorialSourceStorage(SourceStorage):
+    def __init__(self):
+        # import database api here
+
+        # set up database
+        self._storage: dict[int, list[Source]] = {}
+
+    def store(self, documents: list[Document], chat_id: int) -> None:
+        self._storage[chat_id] = [
+            Source(
+                document=document,
+            )
+            for document in documents
+        ]
+
+    def retrieve(
+        self, documents: list[Document], prompt: str, *, chat_id: int
+    ) -> list[Source]:
+        return self._storage[chat_id]
+
+
+# %%
+# [`SourceStorage`][ragna.core.SourceStorage] has two abstract methods,
+# [`store()`][ragna.core.SourceStorage.store] and [`retrieve()`][ragna.core.SourceStorage.retrieve].
+
+# %%
+# [`store()`][ragna.core.SourceStorage.store] takes a list of [`Source`][ragna.core.Source]s
+# as an argument and places them in the database that you are using to hold them. This is
+# different for each different source storage implementation.
+
+# %%
+# [`retrieve()`][ragna.core.SourceStorage.retrieve] returns sources matching
+# the given prompt in order of relevance.
+
+# %%
+# ## Including External Python Objects
+
+# %%
+# If the module containing external object you want to include is not in your
+# [`PYTHONPATH`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH),
+# suppose it is located at the path `~/tutorials/tutorial.py`. You can add `~/tutorial/` to your
+# [`PYTHONPATH`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH) using
+# the command
+# ```bash
+# $ export PYTHONPATH=$PYTHONPATH:~/tutorials/
+# ```
+
+# %%
+# If the external object is already in your
+# [`PYTHONPATH`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH), e.g.
+# already in your virtual environment, then the above `#!bash export` command is not necessary.
+
+# %%
+# Once the module(s) containing your objects is(are) in your
+# [`PYTHONPATH`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH), you can
+# you can include them in Ragna by setting their corresponding
+# [environment variables](../../references/config.md#environment-variables):
```

Comment:
While correct, the regular way is for users to add the components to their configuration file, e.g. https://ragna.chat/en/stable/references/config/#referencing-python-objects.